dialogue-core/src/main/java/com/palantir/dialogue/core/CautiousIncreaseAggressiveDecreaseConcurrencyLimiter.java

```diff
@@ -76,7 +82,59 @@
         return Optional.empty();
     }
 
-    final class Permit implements FutureCallback<Response> {
+    enum Behavior {
+        HOST_LEVEL() {
+            @Override
+            void onSuccess(Response result, PermitControl control) {
+                if ((Responses.isQosStatus(result) && !Responses.isTooManyRequests(result))) {
+                    control.dropped();
+                } else if (Responses.isServerError(result) || Responses.isTooManyRequests(result)) {
+                    control.ignore();
+                } else {
+                    control.success();
+                }
+            }
+
+            @Override
+            void onFailure(Throwable throwable, PermitControl control) {
+                if (throwable instanceof IOException) {
+                    control.dropped();
+                } else {
+                    control.ignore();
+                }
+            }
+        },
+        ENDPOINT_LEVEL() {
+            @Override
+            void onSuccess(Response result, PermitControl control) {
+                if (Responses.isTooManyRequests(result) || Responses.isServerError(result)) {
```

Comment:
could we rename these methods to make em slightly more readable (e.g. isServerError to is5xx or isServerError5xx)? I had assumed that isServerError meant ==500 when it's actually _any_ 5xx...

Reply:
Most places call the entire 5xx range "server error" where the 500 status specifically is an _internal_ server error. Definitely some ambiguity in the spec... perhaps we add a `Range` suffix to the method name?